internal/server/routes.go

```diff
@@ -16,114 +16,111 @@ import (
 	"github.com/infrahq/infra/metrics"
 )
 
+// GenerateRoutes constructs a http.Handler for the primary http and https servers.
+// The handler includes gin middleware, API routes, UI routes, and others.
+//
+// As a side effect of building the handler all API endpoints are registered in
+// the global openAPISchema.
+//
+// The order of routes in this function is important! Gin saves a route along
+// with all the middleware that will apply to the route when the
+// Router.{GET,POST,etc} method is called.
 func (s *Server) GenerateRoutes(promRegistry prometheus.Registerer) (*gin.Engine, error) {
+	a := &API{t: s.tel, server: s}
 	router := gin.New()
+	router.NoRoute(a.notFoundHandler)
 
 	router.Use(gin.Recovery())
-	a := &API{
-		t:      s.tel,
-		server: s,
-	}
-
-	a.registerRoutes(router.Group("/"), promRegistry)
-
-	if err := s.registerUIRoutes(router); err != nil {
-		return nil, err
-	}
-
-	return router, nil
-}
-
-func (a *API) registerRoutes(router *gin.RouterGroup, promRegistry prometheus.Registerer) {
-	router.GET("/healthz", a.healthHandler)
+	router.GET("/healthz", healthHandler)
 
+	// This group of middleware will apply to everything, including the UI
 	router.Use(
 		logging.Middleware(),
 		RequestTimeoutMiddleware(),
 	)
 
-	router.GET("/.well-known/jwks.json", DatabaseMiddleware(a.server.db), a.wellKnownJWKsHandler)
-
-	router.Use(
+	// This group of middleware only applies to non-ui routes
+	api := router.Group("/",
 		sentrygin.New(sentrygin.Options{}),
 		metrics.Middleware(promRegistry),
 		DatabaseMiddleware(a.server.db),
 	)
+	api.GET("/.well-known/jwks.json", a.wellKnownJWKsHandler)
```

Comment:
Small changes in functionality here. Previously we applied the database middleware but not metrics or sentry. With this change we apply those extra two middleware.

I think anything that uses the database should be part of our metrics. The sentry middleware is probably safe to apply to anything so I included it to reduce the number of middleware groups we have.

Reply:
Good catch!